### ISSUE TITLE
Static variables are not exported on multiple platforms.

### ISSUE THREAD

**Issue by Jake-Shadle** [rust-lang/rust#99596]
It appears that #88092 has regressed in nightly, The last working nightly we used was `rustc 1.64.0-nightly (d68e7ebc3 2022-07-20)`, `rustc 1.64.0-nightly (62b272d25 2022-07-21)` was the first nightly that stopped working correctly.

I tried this code:

```rust
// lib.rs
#[no_mangle]
#[used]
pub static FOO: u64 = 4242;
```

I expected to see this happen:

`cargo build --release --target wasm32-unknown-unknown && wasm2wat target/wasm32-unknown-unknown/release/missing_symbols.wasm`

In `rustc 1.62.0 (a8314ef7d 2022-06-27)`, this properly exports the `FOO` static.

```txt
(module
  (table (;0;) 1 1 funcref)
  (memory (;0;) 17)
  (global $__stack_pointer (mut i32) (i32.const 1048576))
  (global (;1;) i32 (i32.const 1048576))
  (global (;2;) i32 (i32.const 1048584))
  (global (;3;) i32 (i32.const 1048592))
  (export "memory" (memory 0))
  (export "FOO" (global 1))
  (export "__data_end" (global 2))
  (export "__heap_base" (global 3))
  (data $.rodata (i32.const 1048576) "\92\10\00\00\00\00\00\00"))
```

Instead, this happened:

`cargo +nightly build --release --target wasm32-unknown-unknown && wasm2wat target/wasm32-unknown-unknown/release/missing_symbols.wasm`

In `rustc 1.64.0-nightly (62b272d25 2022-07-21)`, `FOO` is not exported.

```txt
(module
  (table (;0;) 1 1 funcref)
  (memory (;0;) 16)
  (global $__stack_pointer (mut i32) (i32.const 1048576))
  (global (;1;) i32 (i32.const 1048576))
  (global (;2;) i32 (i32.const 1048576))
  (export "memory" (memory 0))
  (export "__data_end" (global 1))
  (export "__heap_base" (global 2)))
```

### Meta

`rustc --version --verbose`:
```
rustc 1.64.0-nightly (62b272d25 2022-07-21)
binary: rustc
commit-hash: 62b272d25c5bb8b6bb8ac73797d82b8b9a1eabda
commit-date: 2022-07-21
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.6
```

**Comment by bjorn3** [comment]
Regression range: https://github.com/rust-lang/rust/compare/d68e7ebc3...62b272d25 I expect this to be caused by https://github.com/rust-lang/rust/pull/93718 cc @thomcc

**Comment by robbert-vdh** [comment]
This also seems to affect at least `x86_64-unknown-linux-gnu`, `x86_64-pc-windows-gnu`, `x86_64-apple-darwin`, and `aarch64-apple-darwin`. Since about three days ago `#[no_mangle]`+`#[used]` static symbols no longer appear in the data section (/other platform's equivalents). With an explicit `#[used(linker)]` it works, but that feature is of course still unstable.

**Comment by thomcc** [comment]
Hm, interesting. That doesn't make much sense at  I suspect this is an issue with how we compile `#[used(linker)]`? That patch just selects between `#[used(linker)]` and `#[used(compiler)]` based on some settings. CC @nikic, who might have more of an idea about why it may be broken.

I don't have time to fix this right now, and it's very non-obvious to me why this would happen, so if it's as broken as @robbert-vdh suggests, the PR should probably be reverted.

**Comment by thomcc** [comment]
The first PR should definitely fix this. The second fixes the bug. I don't really have the ability to dig into this at the moment, but this should unblock either option.

That said, some evidence in favor of reverting is that this may have been fixed in the meantime, perhaps by https://github.com/rust-lang/rust/pull/95604 ? Not sure.

**Comment by thomcc** [comment]
@Jake-Shadle @robbert-vdh Can you confirm that this is fixed in `rustc 1.64.0-nightly (6dbae3ad1 2022-07-25)`? (Or later)

**Comment by Jake-Shadle** [comment]
Confirmed, nightly build went green again, I'll mark this issue as fixed.
